src/progress.rs

```diff
@@ -54,12 +55,99 @@ impl Default for ProgressState {
-struct Configuration {
+/// A Raft internal representation of a Configuration.
+///
+/// This is corollary to a ConfState, but optimized for `contains` calls.
+#[derive(Clone, Debug, Default, PartialEq, Getters)]
+pub struct Configuration {
+    /// The voter set.
+    #[get = "pub"]
     voters: FxHashSet<u64>,
+    /// The learner set.
+    #[get = "pub"]
     learners: FxHashSet<u64>,
 }
 
+impl Configuration {
+    /// Create a new configuration with the given configuration.
+    pub fn new(
+        voters: impl IntoIterator<Item = u64>,
+        learners: impl IntoIterator<Item = u64>,
+    ) -> Self {
+        Self {
+            voters: voters.into_iter().collect(),
+            learners: learners.into_iter().collect(),
+        }
+    }
+}
+
+impl<Iter1, Iter2> From<(Iter1, Iter2)> for Configuration
+where
+    Iter1: IntoIterator<Item = u64>,
+    Iter2: IntoIterator<Item = u64>,
+{
+    fn from((voters, learners): (Iter1, Iter2)) -> Self {
+        Self {
+            voters: voters.into_iter().collect(),
+            learners: learners.into_iter().collect(),
+        }
+    }
+}
+
+impl From<ConfState> for Configuration {
+    fn from(conf_state: ConfState) -> Self {
+        Self {
+            voters: conf_state.get_nodes().iter().cloned().collect(),
+            learners: conf_state.get_learners().iter().cloned().collect(),
+        }
+    }
+}
+
+impl From<Configuration> for ConfState {
+    fn from(conf: Configuration) -> Self {
+        let mut state = ConfState::default();
+        state.set_nodes(conf.voters.iter().cloned().collect());
+        state.set_learners(conf.learners.iter().cloned().collect());
+        state
+    }
+}
+
+impl Configuration {
+    fn with_capacity(voters: usize, learners: usize) -> Self {
+        Self {
+            voters: HashSet::with_capacity_and_hasher(voters, FxBuildHasher::default()),
+            learners: HashSet::with_capacity_and_hasher(learners, FxBuildHasher::default()),
+        }
+    }
+
+    /// Validates that the configuration is not problematic.
+    ///
+    /// Namely:
+    /// * There can be no overlap of voters and learners.
+    /// * There must be at least one voter.
+    pub fn valid(&self) -> Result<()> {
+        if let Some(id) = self.voters.intersection(&self.learners).next() {
+            Err(Error::Exists(*id, "learners"))?;
+        } else if self.voters.is_empty() {
+            Err(Error::ConfigInvalid(
+                "There must be at least one voter.".into(),
+            ))?;
+        }
+        Ok(())
+    }
+
+    fn has_quorum(&self, potential_quorum: &FxHashSet<u64>) -> bool {
+        self.voters.intersection(potential_quorum).count() >= majority(self.voters.len())
+    }
+
+    /// Returns whether or not the given `id` is a member of this configuration.
+    // Allowed to maintain API consistency. See `HashSet::contains(&u64)`.
+    #[allow(clippy::trivially_copy_pass_by_ref)]
+    pub fn contains(&self, id: &u64) -> bool {
```

Comment:
How about pass `u64`?

Reply:
I'm also curious. I would guess it's by-ref because it makes the implementation cleaner, but it's pretty unusual to take a concrete integer by reference, perhaps places some extra mental burden on the caller.

Reply:
It was mostly because `contains()` on a hashset takes a `&u64`, so I was trying to maintain that. I agree though, let's change it.